enhancements/update/update-status-api.md

```diff
@@ -0,0 +1,410 @@
+
+The Status Insights are not expected to be large data structures, but some contain Kubernetes conditions, so they cannot be considered entirely trivial, and there can be a high number of them. The Health Insights are larger because they can carry a longer-form description of the issue, expected to be around a paragraph of text.
+
+For an API this large, we need to consider the possibility of hitting Kubernetes API resource limits. Even with too many long insights, hitting the 1.5MiB limit does not seem plausible under regular operation. We can protect against the risk by setting appropriate bounds and limits on API fields (especially lists and the long-form description field of health insights), and these limits will be part of the contract for insight producers.
+
+The size and complexity also impact the readability of the API for humans. This is further discussed in [Drawbacks][this-drawbacks].
+
+### Drawbacks
+
+The pattern of the CVO directly deploying a non-operator component is unusual in OpenShift. We could introduce an entirely new Cluster Operator to manage the USC, but because the update functionality is so closely tied to the CVO, an additional layer seems excessive and unnecessary. Adding this layer can be considered in the future if the proposed model is problematic.
+
+Placing the cluster inspection logic directly into the USC puts the OTA team in a position where we need to maintain the logic while not being experts in a significant part of the domain (interpreting `MachineConfigPool` and `Node` states). This is fine as long as we eventually move the logic to the Machine Config Operator to serve as an insight producer, which depends on the future architecture of the system.
+
+The API-backed `oc adm upgrade status` will lose the ability to run against older clusters that do not have the API (or against non-TechPreview clusters while the feature is still in TechPreview). The feature was requested to be an API from the start, and the client-based prototype was meant to be a temporary solution.
+
+Because of their size, the instances of the `UpdateStatus` API will likely be overwhelming for humans to read. As discussed in [API Size and Footprint][this-api-size] under [Risks][this-risks], the serialized resources will be a lot of YAML for humans to read, so humans would depend on tooling provided in the OpenShift ecosystem to benefit from the feature. Some technical users will likely dislike this because they expect Kubernetes resources to be human-readable. It is not possible to reduce the API verbosity without losing valuable information, but it should be possible to provide a human-oriented counterpart (maybe `UpdateStatusSummary`) that would contain aggregated information only. Such a counterpart could be a separate controller in the USC.
+
+The API consumers need to be aware of the possibility of `UpdateStatus` API being absent because the Capability is not enabled, and they will need to degrade gracefully.
```

Comment:
What compatibility do we promise for OC, can this eventually go away? Or because this is a capability, it will always be an issue?

Reply:
Yeah, as long as it is a capability, clients will need to handle the absence of the API. We originally thought this would be an issue for web console but they told us it is fine because web console already does resource discovery and can have conditional behavior on resource presence / absence (see 30f68e208a6302f8a863e0bbb1476668729788f8)